service/topology_coordinator.cc

```diff
@@ -67,6 +68,652 @@ future<inet_address> wait_for_ip(raft::server_id id, const raft_address_map& am,
+        if (src_node_info.tablet_count == 0) {
+            push_back_node_candidate.cancel();
+            nodes_by_load.pop_back();
+        }
+    }
+
+    co_await utils::clear_gently(nodes);
+
+    std::copy(other_dc_replicas.begin(), other_dc_replicas.end(), std::back_inserter(new_replicas));
+
+    co_return upsize_result{new_replicas, tablet_reallocation_status::success};
+}
+
+future<tablet_reallocation_result>
+reallocate_tablets_for_new_rf(schema_ptr s, locator::token_metadata_ptr tm,
+    std::unordered_map<sstring, sstring> const& ks_options) {
+
+    auto const& dcs = tm->get_topology().get_datacenters();
+    std::unordered_map<sstring, size_t> new_rf_per_dc;
+    for (auto const& [option, value] : ks_options) {
+        if (option == "replication_factor")
+            continue;
+        new_rf_per_dc[option] = std::stoul(value);
+    }
+    auto rf_it = ks_options.find("replication_factor");
+    if (rf_it != ks_options.cend()) {
+        auto rf = std::stoul(rf_it->second);
+        for (auto const& dc : dcs) {
+            if (!new_rf_per_dc.contains(dc)) {
+                new_rf_per_dc[dc] = rf;
+            }
+        }
+    }
+
+    rtlogger.info("reallocate_tablets_for_new_rf: computed new_rf_per_dc: {}", new_rf_per_dc);
+
+    rtlogger.info("reallocate_tablets_for_new_rf: sorted_tokens: {}", tm->sorted_tokens());
+
+    assert(!tm->sorted_tokens().empty());
+
+    auto table_id = s->id();
+    locator::tablet_map tablets = tm->tablets().get_tablet_map(table_id);
+
+    rtlogger.info("reallocate_tablets_for_new_rf: table {} new_rf_per_dc: {}", table_id, new_rf_per_dc);
+
+    locator::load_sketch load(tm);
+    co_await load.populate();
+
+    auto host_id_to_dc = [&tm](const locator::host_id& ep) -> std::optional<sstring> {
+        auto node = tm->get_topology().find_node(ep);
+        if (node == nullptr) {
+            return std::nullopt;
+        }
+        return node->dc_rack().dc;
+    };
+
+    tablet_reallocation_result::status_map operation_statuses;
+
+    for (auto tb : tablets.tablet_ids()) {
+        const locator::tablet_info& ti = tablets.get_tablet_info(tb);
+        std::unordered_map<sstring, size_t> dc_rep_factor;
+        locator::tablet_replica_set replicas = ti.replicas;
+
+        for (auto&& replica : replicas) {
+            dc_rep_factor[*host_id_to_dc(replica.host)]++;
+        }
+        for (const auto& [dc, rf] : new_rf_per_dc) {
+            // ensure that we have keys for all DCs
+            dc_rep_factor[dc];
+        }
+
+        rtlogger.info("reallocate_tablets_for_new_rf: tablet {} replicas: {}", tb, replicas);
+        rtlogger.info("reallocate_tablets_for_new_rf: tablet {} dc_rep_factor: {}", tb, dc_rep_factor);
+
+        for (auto& [dc, rf] : dc_rep_factor) {
+            co_await coroutine::maybe_yield();
+
+            auto new_rf = new_rf_per_dc[dc];
+
+            // skip DCs that are not in the new RF per DC map
+            if (!new_rf_per_dc.contains(dc))
```

Comment:
Up above you do `new_rf_per_dc[dc]`, and `operator[]` inserts element if it's missing, so this check is always false, isn't it?